### ISSUE TITLE
Minimal shell with matplotlib dependency fails to initialize

### ISSUE THREAD

**Issue by mtn** [nix-community/poetry2nix#129]
A minimal `shell.nix` with `matplotlib` as the only explicit dependency fails ([minimal example](https://github.com/mtn/minimals/tree/bdac6b45a34f552f77eeab8590a2521f07e1c87f/poetry2nix-install-matplotlib)). The core of the error is:

```
Processing /build/kiwisolver-1.2.0
    ERROR: Command errored out with exit status 1:
     command: /nix/store/j3q3f6nmaif7lmqdsp8fg2hxxrd4w1gh-python3-3.7.6/bin/python3.7 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/build/pip-req-build-nnkyge1u/setup.py'"'"'; __file__='"'"'/build/pip-req-build-nnkyge1u/setup.
py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /build/pip-req-build-nnkyge1u/pip-egg-info
         cwd: /build/pip-req-build-nnkyge1u/
    Complete output (38 lines):
    WARNING: The directory '/homeless-shelter/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo,
 you may want sudo's -H flag.
    WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7ffff6018510>: Failed to estab
lish a new connection: [Errno -2] Name or service not known')': /simple/cppy/
    WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7ffff5f0b550>: Failed to estab
lish a new connection: [Errno -2] Name or service not known')': /simple/cppy/
    WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7ffff5f0b790>: Failed to estab
lish a new connection: [Errno -2] Name or service not known')': /simple/cppy/
    WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7ffff5f0b990>: Failed to estab
lish a new connection: [Errno -2] Name or service not known')': /simple/cppy/
    WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7ffff5f0bb90>: Failed to estab
lish a new connection: [Errno -2] Name or service not known')': /simple/cppy/
    ERROR: Could not find a version that satisfies the requirement cppy>=1.1.0 (from versions: none)
    ERROR: No matching distribution found for cppy>=1.1.0
    Traceback (most recent call last):
      File "/nix/store/4rx2rdjxlj47bl5ds5razad6cqp8sc1f-python3.7-setuptools-44.0.0/lib/python3.7/site-packages/setuptools/installer.py", line 128, in fetch_build_egg
        subprocess.check_call(cmd)
      File "/nix/store/j3q3f6nmaif7lmqdsp8fg2hxxrd4w1gh-python3-3.7.6/lib/python3.7/subprocess.py", line 363, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['/nix/store/j3q3f6nmaif7lmqdsp8fg2hxxrd4w1gh-python3-3.7.6/bin/python3.7', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/build/tmpx3r4ph6f', '--quiet', 'cppy>=1.1.0'
]' returned non-zero exit status 1.
```

When `cppy` is [added as an explicit dependency](https://github.com/mtn/minimals/tree/a1abff4b320c33d2bbfef165c24892aa9344c8f0/poetry2nix-install-matplotlib), the shell instantiates without error and matplotlib works as expected. Is this a bug with `poetry2nix`?

**Comment by adisbladis** [comment]
What channel are you building this against? It's likely your channel is too old for your nixpkgs to have `cppy`.

We add `cppy` to `kiwisolver`, but only optionally: https://github.com/nix-community/poetry2nix/blob/00d58a722d9c140abbb493de08ccc6a4716cfc7d/overrides.nix#L245.

Sadly because of missing _build dependencies_ in `poetry.lock` we need to do this via an override and for those cases we need to pick the dependency from nixpkgs, the proper fix would be to add build time dependencies to `poetry.lock` but that's a much bigger issue.

**Comment by mtn** [comment]
20.03 stable, that's probably the issue

> Sadly because of missing build dependencies in poetry.lock we need to do this via an override and for those cases we need to pick the dependency from nixpkgs, the proper fix would be to add build time dependencies to poetry.lock but that's a much bigger issue.

So you mean `cppy` doesn't show up in the generated build dependencies from poetry right? What I'm confused about is that if I install `matplotlib` with `pip`, `cppy` doesn't show up in the environment.

**Comment by adisbladis** [comment]
> So you mean cppy doesn't show up in the generated build dependencies from poetry right? 

Exactly.

> What I'm confused about is that if I install matplotlib with pip, cppy doesn't show up in the environment.

`pip` will use the wheel by default which is prebuilt and doesn't need `cppy`.

**Comment by adisbladis** [comment]
I think what I will do is to remove the `or null` bit, I don't even remember why I did that. At least then you'll get the error message up front and you can start reasoning about how to remedy the issue at evaluation time rather than at build time.

**Comment by adisbladis** [comment]
Closed in 772b272c965167a7473d6622c4fb45a24efbb88f.

At least we'll fail up front with an attribute error.
